ONSdigital/sds
# Batch process

### Motivation and Context
Transaction has been used to write dataset metadata and unit data into the FireStore, but contains a hard limit of maximum size that it cannot process large dataset (with unit data > 20k)
Transaction is replaced by Batch Write to handle large dataset. However as atomicity only exist for each batch but not across batches, if any batch write triggers an error, although should rarely happen, a clean up function is to be triggered to clean up all data residuals from FireStore. After batch write, a unit_data count checking function is used to make sure the batch write process is successful

### What has changed
- Batch write has replaced transaction in writing dataset and unit data to FireStore
- Batch delete is now used in clean up function that delete data residuals if any batch triggers an error
- The count of unit_data is checked with the value in `total_reporting_unit` as a safety net to ensure the batch write process is completed without problem
- Batch delete logic is now used to delete previous version of dataset when the `retain_dataset_firestore` flag is disabled
- A unit_data count checking function is used to make sure batch write is successful. The function count records by batch using cursor to prevent 530 query timeout error on large dataset
- Set the dataset upload function timeout to maximum of 540 seconds to process large dataset. This is the maximum timeout seconds for event triggered function, which will still timeout for larger dataset (~ 300K) and should be handled in another Jira ticket.
- Set the dataset upload function with 512 MiB memory to process large dataset

### How to test?
- Pass unit test and integration test
- Manual test the clean up function by triggering an exception is one of the batches to check if the data residuals (previously written data of the dataset) is being cleaned up successfully
- To test the previous version dataset deletion logic, change the `retain_dataset_firestore` flag value to FALSE in the gcloud command that deploy the new-dataset function, then deploy the function and upload twice of a dataset containing >20k unit data record. Check if the last version of dataset is deleted completely from FireStore.

### Links
https://jira.ons.gov.uk/browse/SDSS-809

### Issues not fixed
- 540 seconds of maximum timeout will halt the publish process of larger dataset (~300K). GCloud deploy command disallow to override this setting even though manual deploy on console can exceed this limit.
